File: java/3160. Find the Number of Distinct Colors Among the Balls.java
class Solution {
    public int[] queryResults(int limit, int[][] queries) {
        Map<Integer, Integer> colourCount = new HashMap<>();
        Map<Integer, Integer> ballColour = new HashMap<>();
        int[] answer = new int[queries.length];
        int index = 0;
        for (int[] q: queries) {
            int ball = q[0];
            int colour = q[1];
            if (!ballColour.containsKey(ball)) {
                colourCount.put(colour, colourCount.getOrDefault(colour, 0) + 1);
            } else {
                int previousColour = ballColour.get(ball);
                colourCount.put(previousColour, colourCount.get(previousColour)-1);
                if (colourCount.get(previousColour) == 0) {
                    colourCount.remove(previousColour);
                }
                colourCount.put(colour, colourCount.getOrDefault(colour, 0) + 1);
            }
            ballColour.put(ball, colour);
            answer[index++] = colourCount.size();
        }
        return answer;
    }
}

File: python/3160. Find the Number of Distinct Colors Among the Balls.py
class Solution:
    def queryResults(self, limit: int, queries: List[List[int]]) -> List[int]:
        colour_count = {}
        ball_colour = {}
        answer = []
        for q in queries:
            ball, colour = q[0], q[1]
            if ball not in ball_colour:
                if colour not in colour_count:
                    colour_count[colour] = 0
                colour_count[colour] += 1
            else:
                previous_colour = ball_colour[ball]
                if previous_colour != colour:
                    if colour_count[previous_colour] == 1:
                        colour_count.pop(previous_colour)
                    else:
                        colour_count[previous_colour] -= 1
                    if colour not in colour_count:
                        colour_count[colour] = 0
                    colour_count[colour] += 1
            ball_colour[ball] = colour
            answer.append(len(colour_count))
        return answer

                
